sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -2189,3 +2189,293 @@ case class ArrayRemove(left: Expression, right: Expression)
+            var i = 0
+            while (i < array.numElements()) {
+              val elem = array.getLong(i)
+              if (!hs.contains(elem)) {
+                resultArray(pos) = elem
+                hs.add(elem)
+                pos += 1
+              }
+              i += 1
+            }
+          })
+          if (ArraySetLike.useGenericArrayData(LongType.defaultSize, resultArray.length)) {
+            new GenericArrayData(resultArray)
+          } else {
+            UnsafeArrayData.fromPrimitiveArray(resultArray)
+          }
+        case _ =>
+          val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
+          val hs = new OpenHashSet[Any]
+          Seq(array1, array2).foreach(array => {
+            var i = 0
+            while (i < array.numElements()) {
+              val elem = array.get(i, elementType)
+              if (!hs.contains(elem)) {
+                if (arrayBuffer.size > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+                  ArraySetLike.throwUnionLengthOverflowException(arrayBuffer.size)
+                }
+                arrayBuffer += elem
+                hs.add(elem)
+              }
+              i += 1
+            }
+          })
+          new GenericArrayData(arrayBuffer)
+      }
+    } else {
+      ArraySetLike.evalUnionContainsNull(array1, array2, elementType,
+        if (elementTypeSupportEquals) null else ordering)
+    }
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val i = ctx.freshName("i")
+    val pos = ctx.freshName("pos")
+    val value = ctx.freshName("value")
+    val size = ctx.freshName("size")
+    val genericArrayData = classOf[GenericArrayData].getName
+    val (postFix, classTag, getter, setter, javaTypeName, arrayBuilder) = if (!cn) {
+      val ptName = CodeGenerator.primitiveTypeName(elementType)
+      elementType match {
+        case ByteType | ShortType | IntegerType | LongType =>
+          val unsafeArray = ctx.freshName("unsafeArray")
+          (if (elementType == LongType) s"$$mcJ$$sp" else s"$$mcI$$sp",
+            s"scala.reflect.ClassTag$$.MODULE$$.$ptName()",
+            s"get$ptName($i)", s"set$ptName($pos, $value)", CodeGenerator.javaType(elementType),
+            s"""
+              |${ctx.createUnsafeArray(unsafeArray, size, elementType, s" $prettyName failed.")}
+              |${ev.value} = $unsafeArray;
+           """.stripMargin)
+        case _ =>
+          val et = ctx.addReferenceObj("elementType", elementType)
+          ("", s"scala.reflect.ClassTag$$.MODULE$$.Object()",
+            s"get($i, $et)", s"update($pos, $value)", "Object",
+            s"${ev.value} = new $genericArrayData(new Object[$size]);")
+      }
+    } else {
+      ("", "", "", "", "", "")
+    }
+
+    val hs = ctx.freshName("hs")
+    nullSafeCodeGen(ctx, ev, (array1, array2) => {
+      if (classTag != "") {
+        val openHashSet = classOf[OpenHashSet[_]].getName
```

Comment:
As you pointed out, this part works only for `containsNull = false`. 
We ensure `classTag != ""` only when `containsNull = false`. Since it is not easy to understand, I will add a comment regarding this assumption.